### PR TITLE
Error state (from RC1.0)

### PR DESCRIPTION
**Pull request by dpkoch**
This pull commit supersedes #155; it just cherry picks the commit from that one so that it's based off of the `RC1.0` branch instead of `documentation`.

### CODE REVIEW

File: src/sensors.c [hunk 1]
```diff
@@ -96,6 +96,8 @@ static bool update_imu(void);
 // function definitions
 void init_sensors(void)
 {
+  // clear the IMU read error
+  _error_state &= ~(ERROR_IMU_NOT_RESPONDING);
```
Comment (dpkoch) [inline src/sensors.c]: Right now it looks like the `ERROR_IMU_NOT_RESPONDING` flag doesn't get set anywhere?

**Comment by superjax** [comment]
I think you're right.  I was actually thinking about this while I was laying in bed sick, and I was thinking that it should probably be a state machine, like you said.  I was thinking also that the status message should get published every time that an armed state, error state or any of the other "flag" members of the status message gets changed.

i agree, that is future work, but I think that's the answer.

I'll incorporate the prevention of arming if we have an error state, and make sure it's being published on the ROS network, but I think that's as far as I'll go with the error state for this release.

File: src/sensors.c (hunk 1 above)
Comment (superjax) [inline src/sensors.c]: I made a mistake, line 254 was supposed to set the error.

Unfortunately, it gets reset in line 100 during the `imu_not_responding_error()` called in line 255.  That means it basically doesn't ever get called.   However, looking forward to our state machine setup, I hope that it gets published at least once so we know that something happened even if it gets fixed almost immediately.

**Comment by superjax** [comment]
I tested this at my desk.  Seems to be working properly.  Requires an update to fcu_io from the `error_state` branch

**Comment by dpkoch** [comment]
Cool, I like this, and I agree with that plan. I seem to be getting a `TIME_GOING_BACKWARDS` error every time, does that happen to you?

**Comment by superjax** [comment]
I actually haven't gotten it once yet.

**Comment by superjax** [comment]
Ah! just got one.

**Comment by superjax** [comment]
This is weird...

image

Off by a single microsecond?  That is straight-up weird.

Well, unless it's really off by 999 microseconds, and the millisecond timer didn't increment.  I guess that would make sense.

**Comment by superjax** [comment]
Okay, I made a change to breezy that puts highest priority on systick interrupts.  I2C are the next highest.  I ran this on my desk for about 30 minutes and saw 4 I2C errors in the first second, but then none after that.  Loop rate seemed stable and I didn't get any `TIME_GOING_BACKWARDS` errors.

**Comment by dpkoch** [comment]
Awesome, nice work! Looks like this error message stuff has already paid off

**Comment by dpkoch** [comment]
One last question. Right now if we do get a `TIME_GOING_BACKWARDS` error, that flag never gets unset and so we could never recover and arm without rebooting. Is that the behavior we want, or should that flag get cleared somewhere?

**Comment by superjax** [comment]
Hmm.  Probably not.  I guess we could reboot to clear the error, but then that seems extreme.  What conditions would cause it to clear?  We could allow some error codes to arm, and others not to.

**Comment by superjax** [comment]
Okay, we are now just clearing that error the next time that the estimator runs properly.  Ready to merge!

**Pull request merged**